### PR TITLE
Don't force synchronization when task view is opened

### LINKED ISSUE #1030: Cache tasks view content
Currently, when the IDE is starting up, the Tasks view needs to query its content from Gradle. For large projects it can be unbearable. Buildship should cache the content and request a refresh only when the user asks for it.

### PR DESCRIPTION
**Pull request by donat**
Fixes #1030

image

This change will not force a sync when the tasks view is opened. The view content will be updated if:
- The user clicks the refresh button on the view's toolbar
- The users sync a project.

The cache being used lives in memory only the view content will not survive an IDE restart event.

### CODE REVIEW

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/internal/view/task/TreeViewerSelectionChangeListener.java [hunk 1]
```diff
@@ -67,6 +70,16 @@ private ImmutableList<IProject> collectProjectNodesToSelect(List<?> selectedNode
                 if (project.isPresent()) {
                     projects.add(project.get());
                 }
+            } else if (selectedNode instanceof FaultyBuildTreeNode) {
+                FaultyBuildTreeNode faultyNode = (FaultyBuildTreeNode) selectedNode;
+                for(IProject p : CorePlugin.workspaceOperations().getAllProjects()) {
+                    java.util.Optional<GradleBuild> build = CorePlugin.internalGradleWorkspace().getBuild(p);
+                    if (build.isPresent()) {
```
Comment (reinsch82) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/internal/view/task/TreeViewerSelectionChangeListener.java:77]: Can we make this:
```
    CorePlugin.internalGradleWorkspace().getBuild(p)
    .ifPresent(build -> {
        if (((InternalGradleBuild)build).getBuildConfig().equals(faultyNode.getBuildConfiguration())) {
          projects.add(p);
        }
      });
```

or maybe even:

```
    CorePlugin.internalGradleWorkspace().getBuild(p)
    .map(InternalGradleBuild::class::cast)
    .filter(build -> build.getBuildConfig().equals(faultyNode.getBuildConfiguration()))
    .ifPresent(projects::add);
```

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/internal/view/task/WorkbenchSelectionListener.java [hunk 2]
```diff
@@ -82,19 +88,38 @@ private void selectProjectsInTree(List<IProject> projects) {
...
                 Optional<IProject> workspaceProject = selectedNode.getWorkspaceProject();
                 if (workspaceProject.isPresent() && projects.contains(workspaceProject.get())) {
                     selection.add(treeItem);
                 }
+            } else if (data instanceof FaultyBuildTreeNode) {
+                FaultyBuildTreeNode node = (FaultyBuildTreeNode) data;
+                for(IProject project : projects) {
+                    java.util.Optional<GradleBuild> build = CorePlugin.internalGradleWorkspace().getBuild(project);
+                    if (build.isPresent() && ((InternalGradleBuild) build.get()).getBuildConfig().equals(node.getBuildConfiguration())) {
+                        selection.add(treeItem);
```
Comment (reinsch82) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/internal/view/task/WorkbenchSelectionListener.java:104]: same as above `ifPresent(...`

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/internal/view/task/TreeViewerSelectionChangeListener.java (hunk 1 above)
Comment (donat) [reply org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/internal/view/task/TreeViewerSelectionChangeListener.java]: 👍 thanks!

**Comment by altrisi** [comment]
The issue mentioned `when the IDE is starting up`. Given the cache added here `doesn't survive a restart event`, should that issue be reopened or should a new one be created for a more persistent cache? (unless I'm missing something, haven't tried this)

**Comment by donat** [comment]
@altrisi would you mind opening a new feature request with your use cases/pain points?

**Pull request merged**